_pythonSummerCourse/07-02-custom-error-warning.md

```diff
@@ -1,18 +1,142 @@
 ---
 title: "Bài 17. Ngoại lệ/cảnh báo"
 permalink: /pythonSummerCourse/week-07-custom-error-warning/
-last_modified_at: 2022-07-19
+last_modified_at: 2022-08-17
 redirect_from:
   - /theme-setup/
 toc: false
 ---
 
-Nội dung kiến thức
-- Ngoại lệ (exception) và cảnh báo (warning) trong Python
-- Gỡ lỗi (debugging)
+## Nội dung kiến thức
 
-Tài liệu tham khảo
+## 1. Ngoại lệ (exception)
+- Python sử dụng một đối tượng đặc biệt gọi là `exceptions` để quản lý lỗi xảy ra trong quá trình thực hiện chương trình.
+- Một lỗi đơn giản là khi bạn chia một số cho 0 và cần phải bắt ngoại lệ đó.
+  ```py
+  >>> print(2/0)
+  Traceback (most recent call last):
+    File "<stdin>", line 1, in <module>
+  ZeroDivisionError: division by zero
+  ```
+
+### 1.1. `Raise`
+- Sử dụng `raise` để bắt một ngoại lệ nếu điều kiện xảy ra.
+- Để bắt một ngoại lệ khi điều kiện chắc chắn xảy ra thì sử dụng `raise` có thể như ví dụ dưới đây.
+  ```py
+  x = -2
+
+  if x < 0:
+    raise Exception("Numbers below zero")
+  ```
+- Khi bạn thực thi code:
+  ```py
+  ...> python test.py
+  Traceback (most recent call last):
+    File "...\test.py", line 4, in <module>
+      raise Exception("Numbers below zero")
+  Exception: Numbers below zero
+  ```
+
+
+### 1.2. `Assert`
+- Sử dụng `assert` khi nhất định điều kiện đó phải đáp ứng.
+
+  ```py
+  def average(scores):
+      assert len(scores) != 0
+      return sum(scores)/len(scores)
+
+  score = []
+  print("Average of score:", average(score))
+  ```
+Output:
+  ```
+  ...
+  AssertionError
+  ```
+
+### 1.3. `try` and `except`
+- Khối `try` and `except` được sử dụng để catch và handle ngoại lệ.
+  ```py
+  try:
+      print(a)
+  except NameError:
+    print("Variable a is not defined")
+  ```
+- Từ khóa `else` định nghĩa cho một khối code để thực thi nếu không có lỗi nào.
+  ```py
+  else:
+    print("Nothing went wrong!")
+  ```
+- Khối `finally` nếu được chỉ định thì sẽ thực thi bất kể khối `try` dù có gây ra lỗi hay không.
+  ```py
+  finally:
+      print("Hello!")
+  ```
+
```

Comment:
Bổ sung thêm một ví dụ thể hiện toàn bộ kiến thức về try/except/else/finally. Ý tưởng là thiết kế nó ở trong một hàm, để đưa vào nhiều kiểu tham số để nó đi vào các trường hợp khác nhau.

Yêu cầu: phải có ít nhất hai lần dùng `except`.

Reply:
Ví dụ trong [link](https://realpython.com/python-exceptions/#:~:text=Look%20at%20the%20following%20code.%20Here%2C%20you%20first%20call%20the%20linux_interaction()%20function%20and%20then%20try%20to%20open%20a%20file%3A) sau.